src/dune_pkg/opam.mli

```diff
@@ -52,6 +66,6 @@ module Summary : sig
 end
 
 val solve_lock_dir :
-     repo_selection:Repo_selection.t
+     repo_selection:Repo_selection.with_sys Repo_selection.t
```

Comment:
`Repo_selection.t` is a pure value created by parsing command line arguments, so unfortunately at this point I can't inject the environment variables, since that would require evaluating a fiber. Therefore the injection has to be done in a later step. Hence why I added `with_sys_env` that (in an admittedly rather ugly way) patches in the env later (where possible). But that means that `Repo_selection.t` with variables and without would be indistinguishable. To alleviate this I introduced a phantom type to force the user to call `with_sys_env` to turn an `empty Repo_selection.t` into a `with_sys Repo_selection.t`.

But a better solution would be more than welcome, because I am rather unhappy with the code as it is now.

Reply:
One thing I would like to preserve is the `--opam-env=pure` behaviour where the solver is run on a completely empty environment for the purpose of reproducable tests. Also I was thinking of at some point letting the user specify variables to set and I think it would be useful to give the option to set variables starting from a totally empty environment.